### ISSUE TITLE
Add support for TypeScript configuration files

### ISSUE THREAD

**Issue by jesusbotella** [stylelint/stylelint#4940]
> What is the problem you're trying to solve?

As many projects are adopting TypeScript as their main language, I think it would be great to have TypeScript support for stylelint configuration file.

That way, developers will be able to import TypeScript files across their codebases to customize rules in stylelint.

> What solution would you like to see?

An option to have a TypeScript file as a valid configuration file along with JavaScript configuration files.

I saw that stylelint uses `cosmiconfig` for importing configuration files and there's a TypeScript file loader for that. So I can open up a PR if you think that TS configuration files are useful for stylelint users.

**Comment by jeddy3** [comment]
@jesusbotella Thanks for the request and for using the template.

Sounds good to me.

> there's a TypeScript file loader for that.

For those interested, [here's the cosmiconfig loader](https://www.npmjs.com/package/@endemolshinegroup/cosmiconfig-typescript-loader).

Unless anyone has an objection, I'll label as good to go?

**Comment by jesusbotella** [comment]
Thank you!

Just in case anyone wants to know what it takes to add the TypeScript loader, here is the implementation that I have been testing in my project: https://github.com/jesusbotella/stylelint/commit/06af34b48a0d4260b2457bc40dde607be5e435b9

**Comment by jeddy3** [comment]
As we have three thumbs-up already, I'll label as ready to implement.

Please consider [contributing](https://stylelint.io/contributing) the enhancement if you have time.

**Comment by hudochenkov** [comment]
Just for information ESLint declined idea to support `.eslintrc.ts` https://github.com/eslint/rfcs/pull/50

**Comment by ybiquitous** [comment]
@hudochenkov Thank you for sharing the valuable thread! After I read the thread, I started wondering if we really need to support the TypeScript configuration file format. 🤔 

In addtion, `peerDependencies` will be [installed automatically](https://blog.npmjs.org/post/617484925547986944/npm-v7-series-introduction) since npm v7, so [my suggestion using `peerDependenciesMeta`](https://github.com/stylelint/stylelint/pull/4947#issuecomment-698369045) in the PR #4947 may not be effective... 😕 

Sorry to the PR author, but shouldn't we rethink this issue?

**Comment by jesusbotella** [comment]
Yeah, I've read the reasons why ESLint project didn't want to support it and seem reasonable. Installing Typescript in every project that uses stylelint is not a good solution.

My point was that the number of projects that use TypeScript nowadays is growing, and their developers might want to include dynamic configuration for stylelint rules, as I do, to use plugins like [`stylelint-scales`](https://github.com/signal-noise/stylelint-scales) or other use cases.

The proposal of having configuration plugins is good, but stylelint doesn't have a way to do that yet, and it might not be worth just to add support for TypeScript configuration files.

What do you think?

**Comment by jesusbotella** [comment]
I have an update regarding TypeScript support in configuration files. While stylelint might not support them (and I think it's the right thing to do), developers can import TypeScript files with `ts-node` (`tsconfig-paths` might be needed too) within `stylelint.config.js`.

Developers can import `ts-node` at the top of the file, and `ts-lint` will register `.ts` extension in node resolvers. That way, `.ts` files will be transpiled with TypeScript first, and Node will import them without issues. I'll include an example below.

I think we can close the pull request that I opened, and leave this snippet here so that anyone can import TypeScript files in their stylelint configuration.

`stylelint.config.js`
```js
require('ts-node/register');

const dynamicStylelintRules = require('./myTypeScriptFile');

module.exports = {
  ...dynamicStylelintRules
}
```

**Comment by ybiquitous** [comment]
@jesusbotella Thank you for providing the snippet.

I agree with closing PR #4947 if we cannot resolve the problem of increasing extra dependencies.

**Comment by jeddy3** [comment]
Thanks for the update (https://github.com/stylelint/stylelint/issues/4940#issuecomment-702069412). I'll close this issue and the pull request.

**Comment by benedictleejh** [comment]
Hi, I apologise for bumping a closed old issue, but as of https://github.com/eslint/eslint/pull/18134, ESLint supports TypeScript configuration files, and this support was marked as stable in https://github.com/eslint/eslint/pull/19266 (released in ESLint 9.18.0).

A newer RFC for TypeScript config file support is found at https://github.com/eslint/rfcs/pull/117.

Would stylelint be willing to consider reopening this issue in light of the developments from ESLint?

**Comment by Mouvedia** [comment]
> This change will most likely require at least one external tool as either a dependency or a peer dependency.

They chose to add `jiti` as an optional peer dependency thanks to the [`peerDependenciesMeta`](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#peerdependenciesmeta) field.
At least one of its features require node >20.
i.e. https://github.com/unjs/jiti#register-global-esm-loader

An alternative could be [`tsx`](https://github.com/privatenumber/tsx) which support node >=18.
Thing may have changed since summer 2024.

**Comment by jnoordsij** [comment]
Another alternative to do type stripping might be using the API that NodeJS provides by itself, which is still experimental but enabled by default in the latest release (https://nodejs.org/en/blog/release/v23.6.0#unflagging---experimental-strip-types).

**Comment by Atulin** [comment]
FWIW, using `stylelint-define-config` never worked for me for autocompletion. Autotomplete works perfectly when used in a Typescript file, but JS with the mess of JSDoc just never does.

**Comment by silverwind** [comment]
FYI, Node has now [enabled](https://nodejs.org/en/blog/release/v22.18.0#type-stripping-is-enabled-by-default) type stripping by default in v22.18.0 as well.

In theory, stylelint could do something like this:

````js
let config;
try {
  config = await Promise.any([
    "stylelint.config.ts",
    "stylelint.config.js",
    "stylelint.config.mts",
    "stylelint.config.mjs",
  ]).map(file => import(file));
} catch {}
````

This will return the first valid config file out of the files and should also be fully backwards-compatible.

On older versions where type stripping is still experimental, users can set `NODE_OPTIONS=--experimental-strip-types` to make it work.

**Comment by silverwind** [comment]
I just tested `stylelint.config.ts` with Node.js 22.18+ and this appears to work already:

```ts
import type {Config} from 'stylelint';

export default {} satisfies Config;
```

So maybe this is already solved?

**Comment by ybiquitous** [comment]
@silverwind Good point. 👍🏼 

Cosmiconfig, which is a config file loader used by Stylelint, now supports TypeScript config files:

- https://github.com/cosmiconfig/cosmiconfig/pull/311

However, I'd like to point out that this Cosmiconfig feature doesn't depend on the Node.js built-in TypeScript support. Here's a verification: ⬇️ 

<details open><summary><code>package.json</code></summary>
<p>

```json
{
  "devDependencies": {
    "stylelint": "^16.23.1",
    "stylelint-config-standard": "^39.0.0"
  }
}
```

</p>
</details> 

<details open><summary><code>stylelint.config.ts</code></summary>
<p>

```ts
export default {
  extends: ["stylelint-config-standard"]
};
```

</p>
</details> 

<details open><summary>Failed ❌</summary>
<p>

```sh-session
$ node -v
v24.7.0

$ npm i
...

$ npx stylelint <<< 'a{}'
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'typescript' imported from /Users/masafumi/tmp/stylelint-play/node_modules/cosmiconfig/dist/loaders.js
    at Object.getPackageJSONURL (node:internal/modules/package_json_reader:268:9)
    at packageResolve (node:internal/modules/esm/resolve:768:81)
    at moduleResolve (node:internal/modules/esm/resolve:858:18)
    at defaultResolve (node:internal/modules/esm/resolve:990:11)
    at #cachedDefaultResolve (node:internal/modules/esm/loader:749:20)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:726:38)
    at ModuleLoader.getModuleJobForImport (node:internal/modules/esm/loader:312:38)
    at onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:677:36)
    at TracingChannel.tracePromise (node:diagnostics_channel:344:14)
    at ModuleLoader.import (node:internal/modules/esm/loader:676:21)
```

</p>
</details> 

<details open><summary>Succeeded ✅</summary>
<p>

```sh-session
$ npm i -D typescript
...

$ npx stylelint <<< 'a{}'

<input css VmUvBu>
  1:2  ✖  Unexpected empty block  block-no-empty

✖ 1 problem (1 error, 0 warnings)
```

</p>
</details> 

Like this, it's required to install `typescript` for `stylelint.config.ts`.

---

So, I think it's sufficient to add `stylelint.config.ts` to this section: ⬇️ 

https://github.com/stylelint/stylelint/blob/cfbe7720b53be99a1e75cbaa01fade21351e43d3/docs/user-guide/configure.md?plain=1#L20-L29

**Comment by silverwind** [comment]
> However, I'd like to point out that this Cosmiconfig feature doesn't depend on the Node.js built-in TypeScript support

It could attempt to use the native node loader if `process.features.typescript` is truthy, and possibly fall back to the existing cosmicconfig mechanism. Long-term I think cosmicconfig needs to go.

**Comment by ybiquitous** [comment]
Handling the flag on our side doesn't seem like a good idea to me. Instead, Cosmiconfig should do so, as you suggest.

I guess the idea might be helpful as a Cosmiconfig issue.

**Comment by silverwind** [comment]
I opened https://github.com/cosmiconfig/cosmiconfig/issues/350, but am not sure it will be actioned. Currently cosmicconfig compiles the TS and writes a temporary `.mjs` file to the working directory, which is bad of course and will fail on read-only filesystems, for example.

**Comment by benedictleejh** [comment]
I have been testing using a TypeScript config file for a bit since I found out cosmiconfig supported TypeScript by default. In the absence of directly pointing stylelint to `stylelint.config.ts` via the `--config` flag, the automatic search occasionally fails to find the config file, and fails with the error `ConfigurationError: No configuration provided for <file>`. I have not had the time to dig too deeply into the cause, and it being intermittent makes it that much harder to debug. My suspicion is this is related to cosmiconfig compiling to a temporary `.mjs`, and some race condition is involved in the process.

**Comment by silverwind** [comment]
> the automatic search occasionally fails to find the config file, and fails with the error `ConfigurationError: No configuration provided for <file>`

I tested this by running stylelint in a loop with a typescript config and can reproduce as well. Out of 163 stylelint runs, 3 failed with that error. So I think its safe to say typescript configs are not stable.

**Comment by silverwind** [comment]
I noticed that even with `--config stylelint.config.ts` passed, it intermittently fails with `Error: ENOENT: no such file or directory, lstat 'stylelint.config.mjs'`, so it appears there's no workaround for these bugs.

**Comment by silverwind** [comment]
I've opened https://github.com/stylelint/stylelint/issues/8893 for above bug. I think this issue can be closed as support for TS configs is now present.

**Comment by karlhorky** [comment]
@ybiquitous I think this issue was supposed to be closed when the following PR was merged, but that did not happen for whatever reason: 

- https://github.com/stylelint/stylelint/pull/8908

I guess it can be closed now?

**Comment by jeddy3** [comment]
Thanks for spotting this @karlhorky. Closing the issue. 

Let's track the race condition in https://github.com/stylelint/stylelint/issues/8893.
